// @nx/vue/src/generators/storybook-configuration/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.storybookConfigurationGeneratorInternal = exports.storybookConfigurationGenerator = void 0;
const stories_1 = require("../stories/stories");
const devkit_1 = require("@nx/devkit");
const versions_1 = require("../../utils/versions");
async function generateStories(host, schema) {
    await (0, stories_1.default)(host, {
        project: schema.project,
        js: schema.js,
        ignorePaths: schema.ignorePaths,
        skipFormat: true,
        interactionTests: schema.interactionTests ?? true,
    });
}
function storybookConfigurationGenerator(host, schema) {
    return storybookConfigurationGeneratorInternal(host, {
        addPlugin: false,
        ...schema,
    });
}
exports.storybookConfigurationGenerator = storybookConfigurationGenerator;
async function storybookConfigurationGeneratorInternal(host, schema) {
    const { configurationGenerator } = (0, devkit_1.ensurePackage)('@nx/storybook', versions_1.nxVersion);
    const installTask = await configurationGenerator(host, {
        project: schema.project,
        js: schema.js,
        linter: schema.linter,
        tsConfiguration: schema.tsConfiguration ?? true, // default is true
        interactionTests: schema.interactionTests ?? true, // default is true
        configureStaticServe: schema.configureStaticServe,
        uiFramework: '@storybook/vue3-vite',
        skipFormat: true,
        addPlugin: schema.addPlugin,
    });
    if (schema.generateStories) {
        await generateStories(host, schema);
    }
    await (0, devkit_1.formatFiles)(host);
    return installTask;
}
exports.storybookConfigurationGeneratorInternal = storybookConfigurationGeneratorInternal;
exports.default = storybookConfigurationGenerator;
