// @nx/vue/src/generators/application/lib/add-e2e.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addE2e = void 0;
const devkit_1 = require("@nx/devkit");
const web_1 = require("@nx/web");
const versions_1 = require("../../../utils/versions");
async function addE2e(tree, options) {
    switch (options.e2eTestRunner) {
        case 'cypress': {
            const nxJson = (0, devkit_1.readNxJson)(tree);
            const hasPlugin = nxJson.plugins?.some((p) => typeof p === 'string'
                ? p === '@nx/vite/plugin'
                : p.plugin === '@nx/vite/plugin');
            if (!hasPlugin) {
                await (0, web_1.webStaticServeGenerator)(tree, {
                    buildTarget: `${options.projectName}:build`,
                    targetName: 'serve-static',
                    spa: true,
                });
            }
            const { configurationGenerator } = (0, devkit_1.ensurePackage)('@nx/cypress', versions_1.nxVersion);
            (0, devkit_1.addProjectConfiguration)(tree, options.e2eProjectName, {
                projectType: 'application',
                root: options.e2eProjectRoot,
                sourceRoot: (0, devkit_1.joinPathFragments)(options.e2eProjectRoot, 'src'),
                targets: {},
                tags: [],
                implicitDependencies: [options.projectName],
            });
            return await configurationGenerator(tree, {
                ...options,
                project: options.e2eProjectName,
                directory: 'src',
                bundler: 'vite',
                skipFormat: true,
                devServerTarget: `${options.projectName}:serve`,
                baseUrl: 'http://localhost:4200',
                jsx: true,
            });
        }
        case 'playwright': {
            const { configurationGenerator } = (0, devkit_1.ensurePackage)('@nx/playwright', versions_1.nxVersion);
            (0, devkit_1.addProjectConfiguration)(tree, options.e2eProjectName, {
                projectType: 'application',
                root: options.e2eProjectRoot,
                sourceRoot: (0, devkit_1.joinPathFragments)(options.e2eProjectRoot, 'src'),
                targets: {},
                implicitDependencies: [options.projectName],
            });
            return configurationGenerator(tree, {
                ...options,
                project: options.e2eProjectName,
                skipFormat: true,
                skipPackageJson: options.skipPackageJson,
                directory: 'src',
                js: false,
                linter: options.linter,
                setParserOptionsProject: options.setParserOptionsProject,
                webServerCommand: `${(0, devkit_1.getPackageManagerCommand)().exec} nx serve ${options.name}`,
                webServerAddress: 'http://localhost:4200',
            });
        }
        case 'none':
        default:
            return () => { };
    }
}
exports.addE2e = addE2e;
