// @nx/vue/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.vueInitGenerator = exports.setupTailwindGenerator = exports.storiesGenerator = exports.storybookConfigurationGenerator = exports.componentGenerator = exports.libraryGenerator = exports.applicationGenerator = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./src/utils/versions"), exports);
var application_1 = require("./src/generators/application/application");
Object.defineProperty(exports, "applicationGenerator", { enumerable: true, get: function () { return application_1.applicationGenerator; } });
var library_1 = require("./src/generators/library/library");
Object.defineProperty(exports, "libraryGenerator", { enumerable: true, get: function () { return library_1.libraryGenerator; } });
var component_1 = require("./src/generators/component/component");
Object.defineProperty(exports, "componentGenerator", { enumerable: true, get: function () { return component_1.componentGenerator; } });
var configuration_1 = require("./src/generators/storybook-configuration/configuration");
Object.defineProperty(exports, "storybookConfigurationGenerator", { enumerable: true, get: function () { return configuration_1.storybookConfigurationGenerator; } });
var stories_1 = require("./src/generators/stories/stories");
Object.defineProperty(exports, "storiesGenerator", { enumerable: true, get: function () { return stories_1.storiesGenerator; } });
var setup_tailwind_1 = require("./src/generators/setup-tailwind/setup-tailwind");
Object.defineProperty(exports, "setupTailwindGenerator", { enumerable: true, get: function () { return setup_tailwind_1.setupTailwindGenerator; } });
var init_1 = require("./src/generators/init/init");
Object.defineProperty(exports, "vueInitGenerator", { enumerable: true, get: function () { return init_1.vueInitGenerator; } });
tslib_1.__exportStar(require("./src/utils/versions"), exports);
tslib_1.__exportStar(require("./src/utils/add-linting"), exports);
