// @nx/vue/src/utils/create-ts-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractTsConfigBase = exports.createTsConfig = void 0;
const devkit_1 = require("@nx/devkit");
const shared = require("@nx/js/src/utils/typescript/create-ts-config");
function createTsConfig(host, projectRoot, type, options, relativePathToRootTsConfig) {
    const json = {
        compilerOptions: {
            allowJs: true,
            esModuleInterop: false,
            allowSyntheticDefaultImports: true,
            strict: options.strict,
            jsx: 'preserve',
            jsxImportSource: 'vue',
            moduleResolution: 'node',
            resolveJsonModule: true,
        },
        files: [],
        include: [],
        references: [
            {
                path: type === 'app' ? './tsconfig.app.json' : './tsconfig.lib.json',
            },
        ],
    };
    if (options.unitTestRunner === 'vitest') {
        json.references.push({
            path: './tsconfig.spec.json',
        });
    }
    // inline tsconfig.base.json into the project
    if (options.rootProject) {
        json.compileOnSave = false;
        json.compilerOptions = {
            ...shared.tsConfigBaseOptions,
            ...json.compilerOptions,
        };
        json.exclude = ['node_modules', 'tmp'];
    }
    else {
        json.extends = relativePathToRootTsConfig;
    }
    (0, devkit_1.writeJson)(host, `${projectRoot}/tsconfig.json`, json);
    const tsconfigProjectPath = `${projectRoot}/tsconfig.${type}.json`;
    if (host.exists(tsconfigProjectPath)) {
        (0, devkit_1.updateJson)(host, tsconfigProjectPath, (json) => {
            json.compilerOptions ??= {};
            const types = new Set(json.compilerOptions.types ?? []);
            types.add('vite/client');
            json.compilerOptions.types = Array.from(types);
            return json;
        });
    }
    else {
    }
}
exports.createTsConfig = createTsConfig;
function extractTsConfigBase(host) {
    shared.extractTsConfigBase(host);
    if (host.exists('vite.config.ts')) {
        const vite = host.read('vite.config.ts').toString();
        host.write('vite.config.ts', vite.replace(`projects: []`, `projects: ['tsconfig.base.json']`));
    }
}
exports.extractTsConfigBase = extractTsConfigBase;
