// @nx/vue/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lessVersion = exports.sassVersion = exports.autoprefixerVersion = exports.tailwindcssVersion = exports.postcssVersion = exports.eslintPluginVueVersion = exports.vueEslintConfigTypescriptVersion = exports.vueEslintConfigPrettierVersion = exports.vitePluginVueVersion = exports.vueTestUtilsVersion = exports.vueRouterVersion = exports.vueTscVersion = exports.vueVersion = exports.nxVersion = void 0;
exports.nxVersion = require('../../package.json').version;
// vue core
exports.vueVersion = '^3.3.4';
exports.vueTscVersion = '^1.8.8';
exports.vueRouterVersion = '^4.2.4';
// test deps
exports.vueTestUtilsVersion = '^2.4.1';
exports.vitePluginVueVersion = '^4.5.0';
// linting deps
exports.vueEslintConfigPrettierVersion = '7.1.0';
exports.vueEslintConfigTypescriptVersion = '^11.0.3';
exports.eslintPluginVueVersion = '^9.16.1';
// tailwindcss
exports.postcssVersion = '8.4.21';
exports.tailwindcssVersion = '3.2.7';
exports.autoprefixerVersion = '10.4.13';
// other deps
exports.sassVersion = '1.62.1';
exports.lessVersion = '3.12.2';
